static/app/components/tabs/tabList.tsx

```diff
@@ -0,0 +1,189 @@
+import {useContext, useEffect, useMemo, useRef, useState} from 'react';
+import styled from '@emotion/styled';
+import {AriaTabListProps, useTabList} from '@react-aria/tabs';
+import {TabListProps as TabListStateProps, useTabListState} from '@react-stately/tabs';
+import {Orientation} from '@react-types/shared';
+
+import DropdownButton from 'sentry/components/dropdownButton';
+import CompactSelect from 'sentry/components/forms/compactSelect';
+import {IconEllipsis} from 'sentry/icons';
+import {t} from 'sentry/locale';
+import space from 'sentry/styles/space';
+
+import {TabsContext} from './index';
+import {Tab} from './tab';
+
+interface TabListProps<T> extends TabListStateProps<T>, AriaTabListProps<T> {
+  className?: string;
+}
+
+/**
+ * To be used as a direct child of the <Tabs /> component. See example usage
+ * in tabs.stories.js
+ */
+export function TabList<T>({className, ...props}: TabListProps<T>) {
+  const tabListRef = useRef<HTMLUListElement>(null);
+  const [disabledKeys, setDisabledKeys] = useState<React.Key[]>([]);
+  const {rootProps, setTabListState} = useContext(TabsContext);
+  const {
+    orientation,
+    disabled,
+    disabledKeys: _disabledKeys,
+    ...otherRootProps
+  } = rootProps;
+  const ariaProps = useMemo(
+    () => ({
+      isDisabled: disabled,
+      disabledKeys,
+      ...otherRootProps,
+      ...props,
+    }),
+    [disabled, disabledKeys, otherRootProps, props]
+  );
+
+  // Load up list state
+  const state = useTabListState(ariaProps);
+  const {tabListProps} = useTabList({orientation, ...ariaProps}, state, tabListRef);
+  useEffect(() => {
+    setTabListState(state);
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [state.disabledKeys, state.selectedItem, state.selectedKey, props.children]);
+
+  // Dynamically update `disabledKeys` when the tabs list is loaded/updated
+  useEffect(() => {
+    const newDisabledKeys = [...state.collection]
+      .filter(item => {
+        return item.props.disabled;
+      })
+      .map(item => item.key);
+
+    setDisabledKeys(newDisabledKeys);
+  }, [state.collection]);
+
+  // Detect tabs that overflow from the wrapper and put them in an overflow menu
+  const [overflowTabs, setOverflowTabs] = useState<React.Key[]>([]);
+  const tabRefs = useRef<Record<string, React.RefObject<HTMLLIElement>>>({});
+  useEffect(() => {
```

Comment:
nit: might be more readable if this useEffect and `overflowTabs` state was pulled into its own function, like `useOverflowTabs()` or something.

Reply:
great point! did so in https://github.com/getsentry/sentry/pull/38886/commits/d7a4c94f206446bb93e7b378e672fb8d00244d19